### PR TITLE
Pass pointers to events rather than copying them

### PR DESCRIPTION
**Pull request by neilalexander**
This updates the gomatrixserverlib API to deal with pointers to events, to reduce problems where duplicate `Event` objects (when passed by value/through value receivers) would have underlying shared `interface{}` contents. It also should just reduce memory usage a little.

This also fixes a bug or two, like with redacting events and generating event IDs.

Merge this before matrix-org/dendrite#950.

### CODE REVIEW

File: appservice.go [hunk 1]
```diff
@@ -18,5 +18,5 @@ package gomatrixserverlib
 // ApplicationServiceTransaction is the transaction that is sent off to an
 // application service.
 type ApplicationServiceTransaction struct {
-	Events []Event `json:"events"`
+	Events []*Event `json:"events"`
```
Comment (Kegsay) [inline appservice.go]: We don't need a slice of pointers here... it's already passed by reference due to the slice.

File: event.go [hunk 2]
```diff
@@ -311,24 +313,22 @@ func NewEventFromUntrustedJSON(eventJSON []byte, roomVersion RoomVersion) (resul
 				return
 			}
 		}
-
-		eventJSON = redactedJSON
 	}
 
-	result.eventJSON = eventJSON
-
```
Comment (Kegsay) [inline event.go]: Why did you remove these lines?

Comment (neilalexander) [reply event.go]: It's now done at the correct time in `populateFieldsFromJSON`, e.g. after removing `event_id` in the new event format.

File: event.go [hunk 3]
```diff
@@ -357,6 +358,7 @@ func (e *Event) populateFieldsFromJSON(eventJSON []byte) error {
 		if eventJSON, err = sjson.DeleteBytes(eventJSON, "event_id"); err != nil {
 			return err
 		}
+		e.eventJSON = eventJSON
```
Comment (Kegsay) [inline event.go]: Why are we setting this each time for each format and not doing it before the `switch`?

Comment (Kegsay) [reply event.go]: Oh because this version doesn't have an event ID.

File: headeredevent.go [hunk 4]
```diff
@@ -24,23 +24,22 @@ type EventHeader struct {
...
 // Unwrap extracts the event object from the headered event.
-func (e *HeaderedEvent) Unwrap() Event {
+func (e *HeaderedEvent) Unwrap() *Event {
 	if e.RoomVersion == "" {
 		// TODO: Perhaps return an error here instead of panicing
 		panic("gomatrixserverlib: malformed HeaderedEvent doesn't contain room version")
 	}
-	event := e.Event
-	event.roomVersion = e.RoomVersion
-	return event
```
Comment (Kegsay) [inline headeredevent.go]: Why did you change this?

Comment (neilalexander) [reply headeredevent.go]: Because it's no longer taken by value, therefore we can just make sure that the actual event's inner room version is right and return a reference to that directly.

**Comment by neilalexander** [comment]
Superseded by #240.
